File: c++/lattice/gw.cpp
/*******************************************************************************
 *
 * TRIQS: a Toolbox for Research in Interacting Quantum Systems
 *
 * Copyright (C) 2019, The Simons Foundation
 * Authors: H. U.R. Strand
 *
 * TRIQS is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * TRIQS is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * TRIQS. If not, see <http://www.gnu.org/licenses/>.
 *
 ******************************************************************************/

#include "gw.hpp"
#include "../mpi.hpp"

namespace tprf {

chi_wk_t screened_interaction_W(chi_wk_vt PI_wk, chi_k_vt V_k) {

  auto W_wk = make_gf(PI_wk);
  W_wk *= 0.;
  size_t nb = PI_wk.target_shape()[0];

  using scalar_t = chi_wk_t::scalar_t;
  auto I = make_unit_matrix<scalar_t>(nb * nb);

  auto to_matrix = [](auto t) {
    array<scalar_t, 4> t_arr{t, memory_layout_t<4>{0, 1, 2, 3}};
    auto t_mat = make_matrix_view(group_indices_view(t_arr, {0, 1}, {3, 2}));
    return t_mat;
  };

  // MPI and openMP parallell loop
  auto arr = mpi_view(W_wk.mesh());
#pragma omp parallel for
  for (int idx = 0; idx < arr.size(); idx++) {
    auto &[w, k] = arr(idx);

    auto V = to_matrix(V_k[k]);
    auto PI = to_matrix(PI_wk[w, k]);

    // auto W = to_matrix(W_wk[w, k]);

    array<scalar_t, 4> W_arr{nb, nb, nb, nb, memory_layout_t<4>{0, 1, 2, 3}};
    auto W = make_matrix_view(group_indices_view(W_arr, {0, 1}, {3, 2}));

    W = V * inverse(I - PI * V);

    W_wk[w, k] = W_arr;
  }

  W_wk = mpi_all_reduce(W_wk);
  return W_wk;
}

g_wk_t gw_self_energy(chi_wk_vt W_wk, g_wk_vt g_wk) {

  auto g_tr = make_gf_from_fourier<0, 1>(g_wk);
  auto W_tr = make_gf_from_fourier<0, 1>(W_wk);

  auto sigma_tr = make_gf(g_tr);
  sigma_tr *= 0.;

  for (const auto &[t, r] : g_tr.mesh()) {
    for (const auto &[a, b, c, d] : W_tr.target_indices()) {
      sigma_tr[t, r](a, b) += W_tr[t, r](a, b, c, d) * g_tr[t, r](c, d);
    }
  }

  auto sigma_wk = make_gf_from_fourier<0, 1>(sigma_tr);

  return sigma_wk;
}

} // namespace tprf

File: python/triqs_tprf/gw.py
################################################################################
#
# TPRF: Two-Particle Response Function (TPRF) Toolbox for TRIQS
#
# Copyright (C) 2019 by The Simons Foundation
# Copyright (C) 2023 by Hugo U.R. Strand
# Authors: H. U.R. Strand
#
# TPRF is free software: you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation, either version 3 of the License, or (at your option) any later
# version.
#
# TPRF is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# TPRF. If not, see <http://www.gnu.org/licenses/>.
#
################################################################################

import itertools
import numpy as np

# ----------------------------------------------------------------------

import triqs.utility.mpi as mpi

# ----------------------------------------------------------------------

from triqs_tprf.lattice import fourier_wk_to_wr
from triqs_tprf.lattice import fourier_wr_to_wk
from triqs_tprf.lattice import fourier_wr_to_tr
from triqs_tprf.lattice import fourier_tr_to_wr

from triqs_tprf.lattice import chi_wr_from_chi_tr
from triqs_tprf.lattice import chi_tr_from_chi_wr
from triqs_tprf.lattice import chi_wk_from_chi_wr
from triqs_tprf.lattice import chi_wr_from_chi_wk

from triqs_tprf.lattice import chi0_tr_from_grt_PH

from triqs_tprf.lattice import dynamical_screened_interaction_W
from triqs_tprf.lattice import \
    dynamical_screened_interaction_W_from_generalized_susceptibility

from triqs_tprf.lattice import gw_sigma, g0w_sigma
from triqs_tprf.lattice import lindhard_chi00

from triqs_tprf.OperatorUtils import quartic_tensor_from_operator

# ----------------------------------------------------------------------
def bubble_PI_wk(g_wk):

    r""" Compute the particle-hole bubble from the single particle lattice Green's function

    .. math::
        \Pi_{abcd}(i\omega_n, k) = - 
        \mathcal{F}_{\tau, \mathbf{r} \rightarrow i\omega_n, \mathbf{k}}
        \left\{ G_{d\bar{a}}(\tau, \mathbf{r}) G_{b\bar{c}}(-\tau, -\mathbf{r}) \right\}
    
    Parameters
    ----------

    g_wk : TRIQS Green's function (rank 2) on Matsubara and Brillouinzone meshes
           Single particle lattice Green's function.

    Returns
    -------

    PI_wk : TRIQS Green's function (rank 4) on Matsubara and Brillouinzone meshes
            Particle hole bubble.
    """

    nw = len(g_wk.mesh.components[0]) // 2
    
    g_wr = fourier_wk_to_wr(g_wk)
    g_tr = fourier_wr_to_tr(g_wr)
    del g_wr
    PI_tr = chi0_tr_from_grt_PH(g_tr)
    del g_tr
    PI_wr = chi_wr_from_chi_tr(PI_tr, nw=nw)
    del PI_tr
    PI_wk = chi_wk_from_chi_wr(PI_wr)
    del PI_wr

    return PI_wk


# ----------------------------------------------------------------------
def get_gw_tensor(H_int, fundamental_operators):

    """ Takes a TRIQS operator object and extracts the quartic terms
    and returns the corresponding antisymmetrized quartic tensor in
    vertex index order, i.e., cc+cc+. """
    
    U_abcd = quartic_tensor_from_operator(H_int, fundamental_operators)
    
    # -- Group in c^+cc^+c ( from c^+c^+cc )
    U_abcd = np.ascontiguousarray(np.transpose(U_abcd, (0, 3, 1, 2)))    

    # -- Density-density permutations
    U_abcd = (U_abcd + np.transpose(U_abcd, (2,3,0,1)))

    return U_abcd
